Support transaction in FunctionNamespaceManager

caithagoras opened the pull request:
Highlights of the PR:
- `InMemoryFunctionNamespaceManager` for testing purpose.
  - Requires proper definition for `createFunction`.
  - Requires proper definition for `SqlInvokedRegularFunction`
- Transaction support for `FunctionNamespaceManager`.
  - For this PR, transaction support is limited to the read-only method `#getFunctions`.
- Caching support for `FunctionNamespaceManager<SqlInvokedRegularFunction>`
  - Covers both `#getFunctions` and `#getFunctionMetadata`
```
== NO RELEASE NOTE ==
```

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/Column.java:
```diff
@@ -86,10 +77,10 @@ public Type getType()
     }
 
-    public static Column fromResultSet(ResultSet resultSet)
+    public static Column fromResultSet(TypeManager typeManager, ResultSet resultSet)
```
Since `FunctionManager` now requires `TransactionManager` to construct. We need a more formal construction of `TypeManager` and `FunctionManager` through injection.

caithagoras commented:
This PR is ready for review (except for the last commit, which is missing a test).

rongrong commented:
Please fix tests first. Thanks!

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/BuiltInFunction.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @Override
+    public BoundFunction bind(Signature boundSignature)
```
Who calls this? This API looks unnecessary. Even it is, at the very least this needs to check that the given boundSignature actually does not have variable constraints and is compatible with the original signature.

caithagoras replied:
Removed

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/RoutineCharacteristics.java:
```diff
@@ -0,0 +1,121 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class RoutineCharacteristics
```
If you really want to introduce this class, make this part of `SqlFunction`.

rongrong replied:
determinism and call convention are needed for all queries, and they are boolean. They already exist in SqlFunction. I still don't think it's necessary to introduce this class. They can be rolled into the SqlFunction / SqlInvokedRegularFunction directly.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedFunctionParameter.java:
```diff
@@ -0,0 +1,69 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SqlInvokedFunctionParameter
```
This should just be called `SqlParameter` maybe.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedRegularFunction.java:
```diff
@@ -0,0 +1,186 @@
+import static java.util.stream.Collectors.toList;
+
+public class SqlInvokedRegularFunction
+        implements SqlFunction, BoundFunction
```
This looks fishy. Conceptually a `BoundFunction` should be a `SqlFunction`.

rongrong replied:
Which classes with suffix `Function` are not `SqlFunction`?

rongrong replied:
Maybe it makes more sense to rename `SqlFunction` to `GenericSqlFunction` and `BoundFunction` should be `SqlFunction` and `specialize` should return a `SqlFunction`. That way you unifies both builtin and sql functions and both are first class citizen and the definition is clear.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedRegularFunctionId.java:
```diff
@@ -0,0 +1,82 @@
+    private final List<TypeSignature> argumentTypes;
+
+    @JsonCreator
+    public SqlInvokedRegularFunctionId(
```
Why is this class necessary? Also why is this specific to Sql-invoked regular function? name + parameter types is the combination to resolve any functions not just for Sql-invoked regular functions. Also there's no need to require this to be part of the SqlInvokedRegularFunctionHandle. Each function namespace manager should have freedom in choosing how to identify a physical implementation of a function. Engine does not need to dictate what's in the handle.

caithagoras replied:
Removed for now. I will move it to spi when there are other users.

rongrong commented on pom.xml:
```diff
@@ -138,6 +138,7 @@
         <module>presto-proxy</module>
         <module>presto-kudu</module>
         <module>presto-elasticsearch</module>
+        <module>presto-function-namespace</module>
```
presto-function-namespace-manager?

caithagoras commented:
Cleaned up interfaces related to `FunctionNamespaceManager` including
- Removed method `SqlFunction#bind`
- Renamed `BoundFunction` to `BoundSqlFunction`
- Moved `BuiltInBoundFunction` to `FunctionManager` as a private static inner class.

Cleaned up spi classes for `SqlInvokedRegularFunction`:
- Removed `SqlInvokedRegularFunctionId`, and create a private `FunctionId` class for testing purpose within `InMemoryFunctionNamespaceManager`
- Move `RoutineCharacteristics` into `SqlInvokedRegularFunction`

caithagoras commented:
Added a test in `TestFunctionNamespaceManager` to test caching behavior.

caithagoras commented:
@rongrong Comments addressed.

caithagoras commented:
Test failures have been addressed.

caithagoras commented:
Updates since last comment:
- Kept method `FunctionNamespaceManager#getFunctionHandle`
  - Remove `BoundSqlFunction` and `BoundBuiltInFunction`
- Placed classes related to `SqlInvokedRegularFunction` into a new module `presto-sql-function`.
- Placed transactional and caching `AbstractSqlInvokedFunctionNamespaceManager` into `presto-sql-function`
- Fixed tests

caithagoras commented:
@rongrong This PR is ready for review

caithagoras commented on presto-main/src/main/java/com/facebook/presto/metadata/BuiltInFunctionNamespaceManager.java:
```diff
@@ -664,11 +664,17 @@ public BuiltInFunctionNamespaceManager(
 
     @Override
-    public final synchronized void addFunctions(List<? extends SqlFunction> functions)
+    public synchronized void createFunction(BuiltInFunction function, boolean replace)
```
Moved logic to `registerBuiltInFunction`, and `createFunction` should just throw a `PrestoException` with `NOT_SUPPORTED`.

caithagoras commented on presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/SqlInvokedRegularFunction.java:
```diff
@@ -0,0 +1,149 @@
+        return routineCharacteristics.isCalledOnNullInput();
+    }
+
+    @Override
```
Added `COMMENT` to `CREATE FUNCTION` syntax to use a source for description.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionNamespaceManager.java:
```diff
@@ -13,35 +13,42 @@
-    FunctionHandle getFunctionHandle(QueryId queryId, Signature signature);
+    void rollback(FunctionNamespaceTransactionHandle transactionHandle);
+
+    void createFunction(F function, boolean replace);
```
`createFunction` should probably also take a `FunctionNamespaceTransactionHandle` and support transaction. If a transaction is rolled back, the functions created / updated within that transaction should be discarded. You can do that in a follow up commit / PR though. Just add a TODO is fine for now.

caithagoras replied:
Left a TODO.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionNamespaceManager.java:
```diff
@@ -13,35 +13,42 @@
+
+    void createFunction(F function, boolean replace);
+
+    Collection<F> listFunctions();
```
Full transaction support would mean this should also take a transaction handle. Though practically this one is not that important.

caithagoras replied:
Left a TODO for now.

caithagoras commented:
@rongrong TODOs added.
